### PR TITLE
 ext: nrfx: Reserve PPI channels and groups used by pwm_nrf5_sw driver and Bluetooth controller

### PR DESCRIPTION
**Pull request by anangl**
This patch marks the PPI channels and groups used by the pwm_nrf5_sw driver and the Bluetooth controller as occupied and thus unavailable for allocation through nrfx_ppi.

It also adds a build time assertion that checks if the PPI channels used by the Bluetooth controller do not overlap with those assigned to the pwm_nrf5_sw driver (to replace the comments in this driver that were supposed to warn about this threat but had in fact little chance to be read by users).

### CODE REVIEW

File: ext/hal/nordic/nrfx_glue.h [hunk 1]
```diff
@@ -233,16 +233,39 @@ extern "C" {
...
-#define NRFX_DPPI_CHANNELS_USED  0
+#define NRFX_DPPI_CHANNELS_USED 0
 
 /** @brief Bitmask that defines DPPI groups that are reserved for use outside of the nrfx library. */
-#define NRFX_DPPI_GROUPS_USED    0
+#define NRFX_DPPI_GROUPS_USED   0
 
 /** @brief Bitmask that defines PPI channels that are reserved for use outside of the nrfx library. */
-#define NRFX_PPI_CHANNELS_USED  0
+#define NRFX_PPI_CHANNELS_USED  (NRFX_PPI_CHANNELS_USED_BY_BT_CTLR | \
```
Comment (aescolar) [inline ext/hal/nordic/nrfx_glue.h]: nrfx_glue.h seems to be only included from your own nrfx hal top header, which is not part of the nrf52_bsim hal. So this change will not be visible there.

Comment (anangl) [reply ext/hal/nordic/nrfx_glue.h]: Yes, I noticed. And I'm wondering how to solve this appropriately. Do you think that something like:
```
#if defined(SOC_SERIES_BSIM_NRFXX)
#define NRFX_PPI_CHANNELS_USED_BY_PWM_SW  0
#endif
```
could go? Perhaps with a comment that nrfx_glue.h is not processed when building for the simulator.

Comment (aescolar) [reply ext/hal/nordic/nrfx_glue.h]: where would you place those lines?

Comment (anangl) [reply ext/hal/nordic/nrfx_glue.h]: Right before the assertion, [here](https://github.com/zephyrproject-rtos/zephyr/pull/15684/files#diff-d68c1b8276cada6142f4d495d302bea4R475).

Comment (aescolar) [reply ext/hal/nordic/nrfx_glue.h]: Good enough for me

**Comment by nordic-krch** [comment]
Why not include `radio_nrf5_ppi.h` in `nrfx_glue.h`? Then variable would not need to be created and additionally you could do `STATIC_ASSERT` to check if PWM channels to not overlap with BT channels.

**Comment by nordic-krch** [comment]
i see that `radio_nrf5_ppi.h` includes `nrfx/hal/nrf_ppi.h` but then maybe it would be better to split it into 2 files and keep the configuration in one file and function in the other?

**Comment by anangl** [comment]
> Why not include `radio_nrf5_ppi.h` in `nrfx_glue.h`? Then variable would not need to be created

I took such approach in the beginning, but then I thought it was not a good idea to include internal BT controller files into a top-level nrfx file (`nrfx_glue.h` is included from `nrfx.h`), especially when I would need to include `radio_nrf5.h`, not `radio_nrf5_ppi.h` directly, because of the definition of `EVENT_TIMER`. Thus, I thought such solution with external constants would be safer, although still not very nice.

> and additionally you could do `STATIC_ASSERT` to check if PWM channels to not overlap with BT channels.

I'm not sure I understand what you mean. I check it now with a build time (static) assertion.

**Comment by nordic-krch** [comment]
> I'm not sure I understand what you mean. I check it now with a build time (static) assertion.

Right, i didn't noticed that.

> but then I thought it was not a good idea to include internal BT controller files into a top-level nrfx file

On the other hand, there is an extern variable there (`extern const u32_t z_bt_ctlr_used_nrf_ppi_channels;`) so there is that dependency but slightly less obvious.

**Comment by anangl** [comment]
> > but then I thought it was not a good idea to include internal BT controller files into a top-level nrfx file
> 
> On the other hand, there is an extern variable there (`extern const u32_t z_bt_ctlr_used_nrf_ppi_channels;`) so there is that dependency but slightly less obvious.

My concern about this inclusion was the possibility of name collisions. This could be surprising (and irritating) when, after enabling Bluetooth in your application, one of your modules suddenly failed to compile because it apparently redefined the `EVENT_TIMER` macro. And that module could have nothing to do with Bluetooth, it could be only including some nrfx HAL file (but through `nrfx_glue.h` it would get a lot of these internal BT controller symbols).

**Comment by nordic-krch** [comment]
@anangl right, that could be a surprise. I was mainly concerned about verifying that PWM channels do not overlap with BLE and that is in place.

**Review by nordic-krch — APPROVED** [review]

File: ext/hal/nordic/nrfx_glue.h [hunk 2]
```diff
@@ -233,16 +233,39 @@ extern "C" {
...
+
+#if IS_ENABLED(CONFIG_BT_CTLR)
+extern const u32_t z_bt_ctlr_used_nrf_ppi_channels;
+#define NRFX_PPI_CHANNELS_USED_BY_BT_CTLR   z_bt_ctlr_used_nrf_ppi_channels
+#else
+#define NRFX_PPI_CHANNELS_USED_BY_BT_CTLR   0
+#endif
+
+#if IS_ENABLED(CONFIG_PWM_NRF5_SW)
+#define NRFX_PPI_CHANNELS_USED_BY_PWM_SW \
```
Comment (ioannisg) [inline ext/hal/nordic/nrfx_glue.h]: @anangl does this apply for nRF91, as well?

Comment (anangl) [reply ext/hal/nordic/nrfx_glue.h]: Currently you can't compile the pwm_nrf5_sw driver for nRF91.

Comment (ioannisg) [reply ext/hal/nordic/nrfx_glue.h]: Yes, so i wonder: when we will have PWM for 91, you would need to add the _DPPI_ macros in this ifdef block.
Alternatively,

File: ext/hal/nordic/nrfx_glue.h [hunk 3]
```diff
@@ -233,16 +233,39 @@ extern "C" {
 //------------------------------------------------------------------------------
 
 /** @brief Bitmask that defines DPPI channels that are reserved for use outside of the nrfx library. */
-#define NRFX_DPPI_CHANNELS_USED  0
+#define NRFX_DPPI_CHANNELS_USED 0
```
Comment (ioannisg) [inline ext/hal/nordic/nrfx_glue.h]: @anangl I would like to see if these macros could be conditionally defined, only for 91.

File: ext/hal/nordic/nrfx_glue.h (hunk 1 above)
Comment (ioannisg) [inline ext/hal/nordic/nrfx_glue.h]: ...and these ones only for !nrf91

File: ext/hal/nordic/nrfx_glue.h (hunk 2 above)
Comment (anangl) [reply ext/hal/nordic/nrfx_glue.h]: > Yes, so i wonder: when we will have PWM for 91, you would need to add the _DPPI_ macros in this ifdef block.

Yes, after the pwm_nrf5_sw driver is improved to support nRF91, I will update this part of `nrfx_glue.h`. Actually, I planned to do such update directly in #15287 once this PR is merged.

**Review by ioannisg — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**